crates/pathfinder/src/storage.rs

```diff
@@ -145,6 +146,170 @@ fn migrate_from_0_to_1(transaction: &Transaction) -> anyhow::Result<()> {
+            abi        BLOB,
+            definition BLOB
+        )",
+        [],
+    )?;
+
+    // set this to true to have the contracts be dumped into files
+    let dump_duplicate_contracts = false;
+
+    let mut uniq_contracts = 0u32;
+    let todo: u32 = tx
+        .query_row(
+            "select count(1) from (select definition from contracts_v1 group by definition)",
+            [],
+            |r| r.get(0),
+        )
+        .unwrap();
+
+    let mut keccak256 = Keccak256::new();
+    let mut output = vec![0u8; 64];
+
+    let started_at = std::time::Instant::now();
+
+    let mut duplicates = 0;
+
+    // main body of this migration is to split cotracts table into two: contracts and
+    // contracts_code *while* taking care of bug which had mixed up abi and bytecode columns.
+    // the two "faster to access" columns are recreated from the definition.
+    {
+        let mut stmt = tx.prepare("select distinct definition from contracts_v1")?;
+        let mut rows = stmt.query([])?;
+
+        let mut exists = tx.prepare("select 1 from contract_code where hash = ?")?;
+
+        while let Some(r) = rows.next()? {
+            let definition = r.get_ref_unwrap(0).as_blob()?;
+            let raw_definition = zstd::decode_all(definition)?;
+            let (abi, code, hash) =
+                crate::state::contract_hash::extract_abi_code_hash(&raw_definition).with_context(
+                    || format!("Failed to process {} bytes of definition", definition.len()),
+                )?;
+
+            if exists.exists([&hash.to_be_bytes()[..]])? {
+                if dump_duplicate_contracts {
+                    // exists already, this could be a problem
+
+                    keccak256.update(definition);
+                    let cid = <[u8; 32]>::from(keccak256.finalize_reset());
+
+                    hex::encode_to_slice(&cid[..], &mut output[..]).unwrap();
+
+                    let name = std::str::from_utf8(&output[..]).unwrap();
+
+                    let path = format!("duplicate-{:x}-{}.json.zst", hash, name);
+
+                    std::fs::write(path, definition).unwrap();
+                }
+                duplicates += 1;
+            } else {
+                crate::storage::ContractCodeTable::insert(
+                    tx,
+                    crate::core::ContractHash(hash),
+                    &abi,
+                    &code,
+                    &raw_definition,
+                )?;
+                uniq_contracts += 1;
+            }
+
+            let div = 100;
+            if uniq_contracts > 0 && uniq_contracts % div == 0 {
+                let per_one_from_start = started_at.elapsed() / uniq_contracts;
+
+                println!(
+                    "{} more contracts ready, {} to go {:?}, {} duplicates",
+                    div,
+                    todo - uniq_contracts,
+                    (todo - uniq_contracts) * per_one_from_start,
+                    duplicates
+                );
```

Comment:
Would we replace these with debug logs or do you want them in regardless? Or is it because the user will never have this database version so it won't trigger?

Reply:
If the user would have them, then it would take up to 3min to run this migration, which is worrysome when you are the one doing the migration, thinking of backups. It calms the mind to see these prints, however ugly they might be.